Libplanet/Net/Swarm.cs

```diff
@@ -1656,107 +1647,129 @@ CancellationToken cancellationToken
-                        // after actions in stale blocks are unrendered,
-                        // given the `render: false` option here.
-                        workspace.Append(
-                            block,
-                            DateTimeOffset.UtcNow,
-                            evaluateActions: evaluateActions,
-                            renderActions: false
-                        );
-                        receivedBlockCount++;
-                        progress?.Report(new BlockDownloadState
-                        {
-                            TotalBlockCount = totalBlockCount,
-                            ReceivedBlockCount = receivedBlockCount,
-                            ReceivedBlockHash = block.Hash,
-                        });
-                        _logger.Debug($"Block[{block.Hash}] is appended.");
-                    },
-                    cancellationToken);
+                    totalBlockCount = Math.Max(totalBlockCount, receivedBlockCount + hashCount);
+
+                    await GetBlocksAsync(peer, hashesAsArray)
+                        .ForEachAsync(
+                            block =>
+                            {
+                                _logger.Debug(
+                                    $"Trying to append block[{block.Hash}]...");
+
+                                cancellationToken.ThrowIfCancellationRequested();
+
+                                // As actions in this block should be rendered
+                                // after actions in stale blocks are unrendered,
+                                // given the `render: false` option here.
+                                workspace.Append(
+                                    block,
+                                    DateTimeOffset.UtcNow,
+                                    evaluateActions: evaluateActions,
+                                    renderActions: false
+                                );
+                                receivedBlockCount++;
+                                progress?.Report(new BlockDownloadState
+                                {
+                                    TotalBlockCount = totalBlockCount,
+                                    ReceivedBlockCount = receivedBlockCount,
+                                    ReceivedBlockHash = block.Hash,
+                                });
+                                _logger.Debug($"Block[{block.Hash}] is appended.");
+                            },
+                            cancellationToken);
+                }
+            }
+            catch
+            {
+                if (!(workspace is null))
+                {
+                    IStore store = blockChain.Store;
+                    store.DeleteChainId(workspace.Id);
+                }
+            }
+            finally
+            {
+                IStore store = blockChain.Store;
+                foreach (var id in scope.Where(guid => guid != workspace.Id))
```

Comment:
How about inserting an `if` statement in the `foreach` loop, instead of `Where()` method?

Reply:
Rider recommended to use this.